File: app/src/main/java/org/tvheadend/tvhclient/di/modules/RepositoryModule.java
package org.tvheadend.tvhclient.di.modules;

import android.content.Context;

import androidx.annotation.NonNull;

import org.tvheadend.tvhclient.data.db.AppRoomDatabase;
import org.tvheadend.tvhclient.data.repository.AppRepository;
import org.tvheadend.tvhclient.domain.repository.data_source.ChannelData;
import org.tvheadend.tvhclient.domain.repository.data_source.ChannelTagData;
import org.tvheadend.tvhclient.domain.repository.data_source.ConnectionData;
import org.tvheadend.tvhclient.domain.repository.data_source.InputData;
import org.tvheadend.tvhclient.domain.repository.data_source.MiscData;
import org.tvheadend.tvhclient.domain.repository.data_source.ProgramData;
import org.tvheadend.tvhclient.domain.repository.data_source.RecordingData;
import org.tvheadend.tvhclient.domain.repository.data_source.SeriesRecordingData;
import org.tvheadend.tvhclient.domain.repository.data_source.ServerProfileData;
import org.tvheadend.tvhclient.domain.repository.data_source.ServerStatusData;
import org.tvheadend.tvhclient.domain.repository.data_source.SubscriptionData;
import org.tvheadend.tvhclient.domain.repository.data_source.TagAndChannelData;
import org.tvheadend.tvhclient.domain.repository.data_source.TimerRecordingData;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class RepositoryModule {

    private final AppRoomDatabase appRoomDatabase;

    public RepositoryModule(Context context) {
        appRoomDatabase = AppRoomDatabase.Companion.getInstance(context);
    }

    @Singleton
    @NonNull
    @Provides
    AppRoomDatabase providesAppRoomDatabase() {
        return appRoomDatabase;
    }

    @Singleton
    @NonNull
    @Provides
    AppRepository providesAppRepository(AppRoomDatabase db) {
        return new AppRepository(
                new ChannelData(db),
                new ProgramData(db),
                new RecordingData(db),
                new SeriesRecordingData(db),
                new TimerRecordingData(db),
                new ConnectionData(db),
                new ChannelTagData(db),
                new ServerStatusData(db),
                new ServerProfileData(db),
                new TagAndChannelData(db),
                new MiscData(db),
                new SubscriptionData(db),
                new InputData(db));
    }
}

File: data/src/main/java/org/tvheadend/data/di/RepositoryModule.kt
package org.tvheadend.data.di

import android.content.Context
import dagger.Module
import dagger.Provides
import org.tvheadend.data.AppRepository
import org.tvheadend.data.db.AppRoomDatabase
import org.tvheadend.data.source.*
import javax.inject.Singleton

@Module
class RepositoryModule(private val appContext: Context) {

    @Singleton
    @Provides
    fun providesAppRoomDatabase(): AppRoomDatabase {
        return AppRoomDatabase.getInstance(appContext)!!
    }

    @Singleton
    @Provides
    fun providesAppRepository(db: AppRoomDatabase): AppRepository {
        return AppRepository(
                ChannelDataSource(db),
                ProgramDataSource(db),
                RecordingDataSource(db),
                SeriesRecordingDataSource(db),
                TimerRecordingDataSource(db),
                ConnectionDataSource(db),
                ChannelTagDataSource(db),
                ServerStatusDataSource(db),
                ServerProfileDataSource(db),
                TagAndChannelDataSource(db),
                MiscDataSource(db),
                SubscriptionDataSource(db),
                InputDataSource(db))
    }
}
